osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/MigratePlaylistScoresToSoloScoresCommand.cs

```diff
@@ -0,0 +1,190 @@
+                    break;
+
+                using (var db = Queue.GetDatabaseConnection())
+                {
+                    var playlistItems = await db.QueryAsync<MultiplayerPlaylistItem>("SELECT * FROM multiplayer_playlist_items WHERE room_id = @PlaylistId", new
+                    {
+                        PlaylistId = int.Parse(id),
+                    });
+
+                    foreach (var item in playlistItems)
+                    {
+                        if (cancellationToken.IsCancellationRequested)
+                            break;
+
+                        MultiplayerScore[] scores = (await db.QueryAsync<MultiplayerScore>($"SELECT * FROM multiplayer_scores WHERE playlist_item_id = {item.id}")).ToArray();
+
+                        foreach (var score in scores)
+                        {
+                            Console.WriteLine($"Reprocessing score {score.id} from playlist {item.id}");
+
+                            Ruleset ruleset = LegacyRulesetHelper.GetRulesetFromLegacyId(item.ruleset_id);
+                            HitResult maxRulesetJudgement = ruleset.GetHitResults().First().result;
+                            Dictionary<HitResult, int> statistics = JsonConvert.DeserializeObject<Dictionary<HitResult, int>>(score.statistics)!;
+                            List<HitResult> allHits = statistics
+                                                      .SelectMany(kvp => Enumerable.Repeat(kvp.Key, kvp.Value))
+                                                      .ToList();
+                            var maximumStatistics = new Dictionary<HitResult, int>();
+
+                            foreach (var groupedStats in allHits
+                                                         .Select(r => getMaxJudgementFor(r, maxRulesetJudgement))
+                                                         .GroupBy(r => r))
+                            {
+                                maximumStatistics[groupedStats.Key] = groupedStats.Count();
+                            }
+
+                            APIMod[] roomMods = JsonConvert.DeserializeObject<APIMod[]>(item.required_mods)!;
+                            APIMod[] scoreMods = JsonConvert.DeserializeObject<APIMod[]>(score.mods)!;
+
+                            foreach (var m in roomMods)
+                                Debug.Assert(scoreMods.Contains(m));
+
+                            Console.WriteLine($"Mods: {string.Join(',', scoreMods.Select(m => m.ToString()))}");
+
+                            int insertId = await db.InsertAsync(new SoloScore
+                            {
+                                user_id = (int)score.userId,
+                                beatmap_id = (int)score.beatmapId,
+                                ruleset_id = item.ruleset_id,
+                                preserve = true,
+                                ScoreInfo = SoloScoreInfo.ForSubmission(new ScoreInfo
```

Comment:
Since `ForSubmission()` is [mostly intended as a client-side method](https://github.com/ppy/osu/blob/611933a029236fce3ec309f2f9da319f8a0405aa/osu.Game/Online/API/Requests/Responses/SoloScoreInfo.cs#L213-L225), there's some stuff it isn't populating but which could be backpopulated in a server-side context such as this. Looking at other `solo_scores` rows in my DB, fields that could be present in there that I don't believe will be with this implementation are:

- `ended_at`
- `beatmap_id` (arguably redundant)
- `started_at`

Reply:
We probably need to revisit what is being store in the json and what is necessary to be stored in there. I think the original thinking was that the JSON content should stand on its own to represent the play, but I dunno.